pipeline/backend/dummy/dummy.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 Woodpecker Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build integration || test
+// +build integration test
+
+package dummy
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/urfave/cli/v2"
+
+	backend "go.woodpecker-ci.org/woodpecker/v2/pipeline/backend/types"
+)
+
+type dummy struct {
+	kv sync.Map
+}
+
+const (
+	// Step names to control behavior of dummy backend.
+	WorkflowSetupFailUUID = "123456"
+	StepStartFail         = "step_start_fail"
+	StepExecError         = "step_exec_error"
+	EnvKeyStepSleep       = "SLEEP"
+	EnvKeyStepType        = "EXPECT_TYPE"
+
+	// Internal const.
+	stepStateStarted   = "started"
+	stepStateDone      = "done"
+	testServiceTimeout = 1 * time.Second
+)
+
+// New returns a dummy backend.
+func New() backend.Backend {
+	return &dummy{
+		kv: sync.Map{},
+	}
+}
+
+func (e *dummy) Name() string {
+	return "dummy"
+}
+
+func (e *dummy) IsAvailable(_ context.Context) bool {
+	return true
+}
+
+func (e *dummy) Flags() []cli.Flag {
+	return nil
+}
+
+// Load new client for Docker Backend using environment variables.
+func (e *dummy) Load(_ context.Context) (*backend.BackendInfo, error) {
+	return &backend.BackendInfo{
+		Platform: "dummy",
+	}, nil
+}
+
+func (e *dummy) SetupWorkflow(_ context.Context, _ *backend.Config, taskUUID string) error {
+	if taskUUID == WorkflowSetupFailUUID {
+		return fmt.Errorf("expected fail to setup workflow")
+	}
+	log.Trace().Str("taskUUID", taskUUID).Msg("create workflow environment")
+	e.kv.Store("task_"+taskUUID, "setup")
+	return nil
+}
+
+func (e *dummy) StartStep(_ context.Context, step *backend.Step, taskUUID string) error {
+	log.Trace().Str("taskUUID", taskUUID).Msgf("start step %s", step.Name)
+
+	// internal state checks
+	_, exist := e.kv.Load("task_" + taskUUID)
+	if !exist {
+		return fmt.Errorf("expect env of workflow %s to exist but found none to destroy", taskUUID)
+	}
+	stepState, stepExist := e.kv.Load(fmt.Sprintf("task_%s_step_%s", taskUUID, step.UUID))
+	if stepExist {
+		// Detect issues like https://github.com/woodpecker-ci/woodpecker/issues/3494
+		return fmt.Errorf("StartStep detected already started step '%s' (%s) in state: %s", step.Name, step.UUID, stepState)
+	}
+
+	if step.Name == StepStartFail {
```

Comment:
```suggestion
	if step.Environment[EnvKeyFailStartStep] {
```

When testing with two steps that should fail using the name is already quite limiting as you don't know which step it is from the logs. Therefore I would like to directly use an env var.